rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+
+## Reviewers
+
+- [ ] TBD
+- [ ] TBD
+
+## Project links
+
+- [Project Thread](https://github.com/WordPress/openverse/issues/377)
+- [Project Proposal](./20230309-project_proposal_detecting_sensitive_textual_content.md)
+- [Milestone (pending)](#)
+- [Implementation Plan: Managing A Sensitive Terms List](https://github.com/WordPress/openverse/pull/911)
+
+Please refer to additional prior art linked in the project thread. The most
+significant technical prior art is this exploratory PR for creating a filtered
+index: https://github.com/WordPress/openverse-api/pull/1108
+
+## Expected outcomes
+
+- Results are filtered efficiently based on a list of sensitive terms.
+- API results include a call-out on the result objects designating individual
+  results that have sensitive textual content.
+
+## Tools and dependencies
+
+We will not need any new tools on the API side for this work.
+
+## Feature flags
+
+We should introduce a single "feature flag" on the API as an undocumented and
+unstable query parameter `unstable__include_sensitive_results`.
+
+## API versioning
+
+There are no API versioning concerns. We already filter sensitive content based
+on provider data, we'll just be expanding upon that existing feature by
+incorporating a new signal into the filter. I do not believe this requires a new
+API version prefix or distinctly new parameter.
+
+## Areas touched
+
+Implementing this plan will require making changes in the following areas:
+
+- API
+- Elasticsearch
+- Ingestion server
+- Airflow data refresh DAG factory
+
+## Terms
+
+- "Origin index": This is the index we have now that includes all documents.
+- "Filtered index": This is the index we will have after this implementation
+  plan is implemented. Its documents are derived from the origin index but
+  exclude any that have sensitive terms in their textual content.
+
+## Overview of implementation process
+
+In depth descriptions of each of these steps are covered below. Each step will
+be implemented as a single pull request including appropriate tests.
+
+1. Update the ingestion server with a new action to use the reindex API to
+   generate a new index based on the origin index that only includes documents
+   that do not have sensitive terms in the three textual content fields
+   (`title`, `description`, `tags.name`). This is the filtered index.
+   - Requires the following:
+     - SENSITIVE_TERMS injected as new environment variable. See
+       [the related IP](https://github.com/WordPress/openverse/pull/911) for how
+       this will be implemented in Terraform.
+     - Ingestion server action to create the new index using `reindex` and an
+       update the promote action to allow it to promote (point alias) for the
+       filtered index
+     - Updates to the `load_sample_data.sh` script to generate the filtered
+       index
+     - A benign list of terms to use for local testing that matches results in
```

Comment:
Yes, good point. Do you think we should maintain the testing list external from the application, rather than as the default value when reading the environment variable?

Reply:
+1, I think that would be the safest approach. Great call out!

Reply:
To clarify, you think that we should _not_ just use the default value for the environment variable?

If we don't then we'd need to update `env.docker` for the ingestion server and copy those to the ingestion server settings.

If we set the default in code then it is (in my opinion) slightly easier to understand in testing. We can restrict the default to only apply not in production:

```py
SENSITIVE_TERMS = config(
    "SENSITIVE_TERMS",
    cast=lambda x: x.split("\n"),
    default=( # just an example list
        """
        water
        dogs
        cactus
        """
        if ENVIRONMENT != "production" else None
    )
)
```

We can raise an error in `create_and_populate_filtered_index` if `SENSITIVE_TERMS` is None.

Reply:
The addition of the environment check works just fine and I agree, easier to understand. Your example here is totally fine with me.